### PR TITLE
 Introduce refresh/revert tests for kernel/snapd/gadget snaps (New)

### PR DESCRIPTION
**Pull request by pieqq**
This branch is exactly the same as the one used for #724, but with a different name in order not to break the runners policy. Please check #724 for explanation, how to test, all the previous comments addressed, etc.

-------------

## Description

Rework @patliuu 's kernel snap refresh/revert tests to

- be more generic (so now the tests support gadget and snapd snaps in addition)
- allow to refresh to a specific revision instead of a channel
- add jobs to not only refresh to/revert from stable, but also from base revision (base revision = the revision of a snap that is present on the image at install time). This is to allow simulating a scenario where a freshly installed device gets updated to the latest snap available.
- create a separate test plan to check all of the snaps for refresh/revert (the jobs will be skipped if the revisions are the same)

Please see individual commit messages for more info.

## Resolved issues

CHECKBOX-717

## Documentation

- Documented the `snap_update_test.py` script (docstrings, type hints)
- Added unit tests for it
- Put additional info in commit messages for later reference

## Tests

- `./manage.py test -u` are passing
- I ran each test plan on a Raspberry Pi 4 running UC22 by sideloading the base provider. Sample submission: https://certification.canonical.com/hardware/202212-30981/submission/332039/ (see submission description for an explanation)

### CODE REVIEW

**Comment by pieqq** [comment]
I run the `com.canonical.certification::ubuntucore-snap-refresh-revert` test plan on the target device in the lab: https://certification.canonical.com/hardware/202207-30448/submission/341140/

Results are as expected:

- `snapd` (beta revision 20515) was refreshed/reverted to and from the base rev (19267) and the stable rev (20290) :heavy_check_mark: 
- `pc-kernel` (beta revision 1466) was refresh/reverted to and from the base rev (1289) but not the stable rev since it's the same as beta rev (1466) :heavy_check_mark: 
- `xinyi-pc-classic` (gadget snap, beta revision 18) was not refreshed/reverted from neither base nor stable since all of the tracks point to the same revision (18) xinyi-pc-classic

**Review by Hook25 — CHANGES_REQUESTED** [review]
I have a couple of notes about tests and possibly a missing arg. Overall a nice contributin. If the old PR is superseeded by this one, please close it.

File: providers/base/tests/test_snap_update_test.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+
+class SnapRefreshRevertTests(unittest.TestCase):
+    def test_snap_refresh_same_revision(self):
+        mock_self = MagicMock()
+        mock_self.revision = "1"
+        mock_snap_info = MagicMock()
+        mock_snap_info.installed_revision = "1"
+        mock_self.snap_info = mock_snap_info
+        logging.disable(logging.CRITICAL)
```
[the remark this answers is not in the document]
Comment (pieqq) [reply providers/base/tests/test_snap_update_test.py:130]: I tried mocking `snap_update_test.logging` by patching it for the affected test methods, but the logs are still showing up when running the tests.

In the end, I went for the `setUpClass/tearDownClass` way to disable and re-enable logging.

File: providers/base/bin/snap_update_test.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+            if current_time - start_time >= timeout:
+                raise SystemExit(
+                    "{} snap {} did not complete within {} seconds".format(
+                        self.name, type, timeout
+                    )
+                )
+            logger.info("Waiting for %s snap %s to be done...",
+                        self.name,
+                        type
+                        )
```
[the remark this answers is not in the document]
Comment (pieqq) [reply providers/base/bin/snap_update_test.py]: Done.

File: providers/base/tests/test_snap_update_test.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        assert "tracking:" in mock_stdout.getvalue()
+        assert "base_rev:" in mock_stdout.getvalue()
+        assert "stable_rev:" in mock_stdout.getvalue()
+        assert "candidate_rev:" in mock_stdout.getvalue()
+        assert "beta_rev:" in mock_stdout.getvalue()
+        assert "edge_rev:" in mock_stdout.getvalue()
+        assert "original_installed_rev:" in mock_stdout.getvalue()
+        # Make sure there is a blank line at the end, as this is required by
+        # Checkbox resource jobs
+        assert mock_stdout.getvalue().endswith("\n\n")
```
[the remark this answers is not in the document]
Comment (pieqq) [reply providers/base/tests/test_snap_update_test.py]: Done

File: providers/base/tests/test_snap_update_test.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        self, mock_save_change_info, mock_snapd_refresh
+    ):
+        mock_self = MagicMock()
+        mock_self.revision = "1"
+        mock_snap_info = MagicMock()
+        mock_snap_info.installed_revision = "2"
+        mock_self.snap_info = mock_snap_info
+        logging.disable(logging.CRITICAL)
+        snap_update_test.SnapRefreshRevert.snap_refresh(mock_self)
+        self.assertTrue(snap_update_test.save_change_info.called)
```
[the remark this answers is not in the document]
Comment (pieqq) [reply providers/base/tests/test_snap_update_test.py]: I modified this test to check that the `Snapd.refresh()` method was called. In my previous implementation, the `mock_self` object passed to `snap_update_test.SnapRefreshRevert.snap_refresh` did not actually initialized `self.snapd` to `Snapd()`, so I added that to be able to check that the `refresh()` method was called.

File: providers/base/tests/test_snap_update_test.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+    @patch("snap_update_test.Snapd.revert")
+    @patch("snap_update_test.load_change_info")
+    @patch("snap_update_test.save_change_info")
+    def test_snap_revert(
+        self, mock_save_change_info, mock_load_change_info, mock_snapd_revert
+    ):
+        mock_self = MagicMock()
+        mock_snap_info = MagicMock()
+        mock_self.snap_info = mock_snap_info
```
[the remark this answers is not in the document]
Comment (pieqq) [reply providers/base/tests/test_snap_update_test.py]: Done.

File: providers/base/bin/snap_update_test.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+        if type == "refresh":
+            tested_rev = data["destination_revision"]
+        else:
+            tested_rev = data["original_revision"]
+        if current_rev != tested_rev:
+            msg = (
+                "Current revision ({}) is different from expected revision "
+                "({})"
+            ).format(current_rev, tested_rev)
+            raise SystemExit()
```
[the remark this answers is not in the document]
Comment (pieqq) [reply providers/base/bin/snap_update_test.py]: Good catch! Fixed in fd02e6f68a785947abaf62856249a69f98b638d9

**Review by Hook25 — APPROVED** [review]
+1, TY for the patience!

**Pull request merged**